### PR TITLE
[TechDocs] Provide a migration path from case-sensitive to lower-case entity triplets in docs storage

### PR DESCRIPTION
**Pull request by iamEAP**
## Hey, I just made a Pull Request!

This is step one of a multi-step process toward resolving #4367.  This adds the necessary logic on a publisher-by-publisher basis to migrate objects in storage from, e.g. `default/Component/foo/index.html` to `default/component/foo/index.html`.

Those steps are:

1. Define and implement migration logic for storage providers. <-- We are here.
2. Expose the implementations in `techdocs-cli`
3. Update publisher implementations to read/write lower-case'd versions of the entity triplet, regardless of how the path arguments come in (expectation being that those running external storage configurations will run the aforementioned migration path prior to deployment)

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts [hunk 1]
```diff
@@ -289,4 +293,54 @@ export class AzureBlobStoragePublish implements PublisherBase {
...
+      this.logger.warn(`Unable to migrate ${originalPath}: ${e.message}`);
+    }
+  }
+
+  async migrateDocsCase({
+    removeOriginal = false,
+    concurrency = 25,
+  }): Promise<void> {
+    const promises = [];
+    const limiter = limiterFactory(concurrency);
```
Comment (soapraj) [inline packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:331]: just curious to why different ways of creating a limiter in here and awsS3?

File: packages/techdocs-common/src/stages/publish/helpers.test.ts [hunk 2]
```diff
@@ -73,3 +77,17 @@ describe('getFileTreeRecursively', () => {
     expect(fileList).toContain(path.resolve(root, 'subDirA/file2'));
   });
 });
+
+describe('lowerCaseEntityTripletInStoragePath', () => {
+  it('returns lower-cased entity triplet path', () => {
```
Comment (soapraj) [inline packages/techdocs-common/src/stages/publish/helpers.test.ts:82]: would it be good to test if you get paths in unexpected formats? like not having a triplet? Eg: /default/component/IMAGE.png ?

File: packages/techdocs-common/src/stages/publish/helpers.ts [hunk 3]
```diff
@@ -82,3 +82,24 @@ export const getFileTreeRecursively = async (
...
+ * Path must not include a starting slash.
+ *
+ * @example
+ * lowerCaseEntityTripletInStoragePath('default/Component/backstage')
+ * // return default/component/backstage
+ */
+export const lowerCaseEntityTripletInStoragePath = (
+  originalPath: string,
+): string => {
+  const [namespace, kind, name, ...parts] = originalPath.split('/');
```
Comment (soapraj) [inline packages/techdocs-common/src/stages/publish/helpers.ts:100]: we may need to check if we have a minimum of 3 parts in the part other than the filename.

**Review by soapraj — APPROVED** [review]

File: packages/techdocs-common/src/stages/publish/helpers.ts (hunk 3 above)
Comment (iamEAP) [reply packages/techdocs-common/src/stages/publish/helpers.ts:100]: Yeah, good call.  Maybe we throw/catch and log a warning in that case (`Encountered file unmanaged by TechDocs ${fileName}. Skipping.`)

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts (hunk 1 above)
Comment (camilaibs) [reply packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:331]: Both are the same default export, but use a different name when importing:
https://github.com/backstage/backstage/blob/34efce9c44ed2fd8414da77089e04f055405d361/packages/techdocs-common/src/stages/publish/awsS3.ts#L24
https://github.com/backstage/backstage/blob/34efce9c44ed2fd8414da77089e04f055405d361/packages/techdocs-common/src/stages/publish/azureBlobStorage.ts#L25

File: packages/techdocs-common/src/stages/publish/helpers.ts (hunk 3 above)
Comment (camilaibs) [reply packages/techdocs-common/src/stages/publish/helpers.ts:108]: Hey @soapraj can you review it again, please?

File: packages/techdocs-common/src/stages/publish/helpers.test.ts (hunk 2 above)
Comment (camilaibs) [reply packages/techdocs-common/src/stages/publish/helpers.test.ts:82]: Hey @soapraj can you review it again, please?

**Review by soapraj — APPROVED** [review]

**Pull request merged**